File: tests/dotnet/DecompressorTest.cs
/*
 * DecompressorTest.cs
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
using System;
using System.Text;
using Xunit;

namespace PDFjet.NET {
/// <summary>The stream filters and predictors that reading a PDF needs, and the compressor.</summary>
public class DecompressorTest {
    private static byte[] Bytes(params int[] values) {
        byte[] result = new byte[values.Length];
        for (int i = 0; i < values.Length; i++) {
            result[i] = (byte) values[i];
        }
        return result;
    }

    private static byte[] Ascii(string text) {
        return Encoding.ASCII.GetBytes(text);
    }

    [Fact]
    public void LzwDecodesTheExampleOfTheStandard() {
        // ISO 32000-1, 7.4.4.2: the encoding of "-----A---B".
        byte[] encoded = Bytes(0x80, 0x0B, 0x60, 0x50, 0x22, 0x0C, 0x0C, 0x85, 0x01);
        Assert.Equal("-----A---B", TestSupport.Latin1(Decompressor.LZWDecode(encoded)));
    }

    [Fact]
    public void AsciiHexDecodeSkipsWhitespaceAndStopsAtTheEndMarker() {
        Assert.Equal("Hello", TestSupport.Latin1(Decompressor.ASCIIHexDecode(Ascii("48 65 6C6C6F>"))));
    }

    [Fact]
    public void AsciiHexDecodeReadsAMissingLastDigitAsZero() {
        Assert.Equal(Bytes(0x70), Decompressor.ASCIIHexDecode(Ascii("7>")));
    }

    [Fact]
    public void Ascii85DecodesWithAndWithoutThePrefix() {
        // base64.a85encode(b"Hello world") in Python.
        Assert.Equal("Hello world", TestSupport.Latin1(Decompressor.ASCII85Decode(Ascii("87cURD]j7BEbo7~>"))));
        Assert.Equal("Hello world", TestSupport.Latin1(Decompressor.ASCII85Decode(Ascii("<~87cURD]j7BEbo7~>"))));
    }

    [Fact]
    public void Ascii85DecodesZAsFourZeroBytesAndAPartialLastGroup() {
        Assert.Equal(Bytes(0, 0, 0, 0), Decompressor.ASCII85Decode(Ascii("z~>")));
        Assert.Equal(Bytes(0, 0, 0, 0, 'a', 'b'), Decompressor.ASCII85Decode(Ascii("z@:B~>")));
    }

    [Fact]
    public void RunLengthDecodeCopiesLiteralsAndRepeatsRuns() {
        byte[] encoded = Bytes(2, 'a', 'b', 'c', 254, 'x', 128);
        Assert.Equal("abcxxx", TestSupport.Latin1(Decompressor.RunLengthDecode(encoded)));
    }

    [Fact]
    public void PngPredictorsUndoEachRowFilter() {
        // Each row is the filter type and three one byte samples.
        Assert.Equal(Bytes(1, 2, 3), Decompressor.ApplyPredictor(Bytes(1, 1, 1, 1), 11, 1, 8, 3));
        Assert.Equal(Bytes(1, 2, 3, 2, 3, 4), Decompressor.ApplyPredictor(Bytes(0, 1, 2, 3, 2, 1, 1, 1), 12, 1, 8, 3));
        Assert.Equal(Bytes(2, 5, 8), Decompressor.ApplyPredictor(Bytes(3, 2, 4, 6), 13, 1, 8, 3));
        Assert.Equal(Bytes(1, 2, 3), Decompressor.ApplyPredictor(Bytes(4, 1, 1, 1), 14, 1, 8, 3));
    }

    [Fact]
    public void TiffPredictorAddsTheSampleToTheLeft() {
        Assert.Equal(Bytes(1, 2, 3, 5, 5, 5), Decompressor.ApplyPredictor(Bytes(1, 1, 1, 5, 0, 0), 2, 1, 8, 3));
    }

    [Fact]
    public void PredictorOneAndInvalidParametersLeaveTheDataAsItIs() {
        byte[] data = Bytes(9, 8, 7);
        Assert.Equal(data, Decompressor.ApplyPredictor(data, 1, 1, 8, 3));
        Assert.Equal(data, Decompressor.ApplyPredictor(data, 12, 0, 8, 3));
    }

    [Fact]
    public void InflateUndoesDeflate() {
        byte[] data = new byte[100000];
        new Random(42).NextBytes(data);
        Array.Fill(data, (byte) 'x', 50000, 40000);
        Assert.Equal(data, Decompressor.Inflate(Compressor.Deflate(data)));
    }

    [Fact]
    public void InflateRejectsATruncatedStream() {
        byte[] deflated = Compressor.Deflate(Ascii("hello hello hello hello"));
        Assert.ThrowsAny<Exception>(() => Decompressor.Inflate(deflated[..6]));
    }

    [Fact]
    public void InflateIgnoresBytesAfterTheEndOfAStream() {
        byte[] data = Ascii("hello hello hello hello");
        byte[] deflated = Compressor.Deflate(data);
        byte[] padded = new byte[deflated.Length + 3];
        deflated.CopyTo(padded, 0);
        padded[deflated.Length] = (byte) '\r';
        padded[deflated.Length + 1] = (byte) '\n';
        padded[deflated.Length + 2] = 0x42;
        Assert.Equal(data, Decompressor.Inflate(padded));
        Assert.Empty(Decompressor.Inflate(Compressor.Deflate(new byte[0])));
    }

    [Fact]
    public void InflateRejectsEveryTruncationOfAStream() {
        byte[] data = new byte[2000];
        new Random(3).NextBytes(data.AsSpan(0, 1000));
        data.AsSpan(1000).Fill((byte) 'a');
        byte[] deflated = Compressor.Deflate(data);
        Assert.Equal(data, Decompressor.Inflate(deflated));
        // The last 4 bytes are the Adler-32 checksum, which is checked too.
        for (int length = 0; length < deflated.Length; length++) {
            byte[] truncated = deflated[..length];
            Assert.ThrowsAny<Exception>(() => Decompressor.Inflate(truncated));
        }
    }

    [Fact]
    public void InflateRejectsAWrongChecksumOrHeader() {
        byte[] deflated = Compressor.Deflate(Encoding.ASCII.GetBytes("PDFjet PDFjet PDFjet"));
        byte[] wrongChecksum = (byte[]) deflated.Clone();
        wrongChecksum[wrongChecksum.Length - 1] ^= 1;
        Assert.ThrowsAny<Exception>(() => Decompressor.Inflate(wrongChecksum));
        byte[] wrongMethod = (byte[]) deflated.Clone();
        wrongMethod[0] = 0x77;     // Method 7, and the check no longer fits
        Assert.ThrowsAny<Exception>(() => Decompressor.Inflate(wrongMethod));
        byte[] wrongCheck = (byte[]) deflated.Clone();
        wrongCheck[1] ^= 1;
        Assert.ThrowsAny<Exception>(() => Decompressor.Inflate(wrongCheck));
    }

    [Fact]
    public void InflatePrefixWithItsBytesNeedsNoChecksum() {
        byte[] data = Encoding.ASCII.GetBytes("PDFjet PDFjet PDFjet");
        byte[] deflated = Compressor.Deflate(data);
        deflated[deflated.Length - 1] ^= 1;
        Assert.Equal(data[..5], Decompressor.InflatePrefix(deflated, 5));
        Assert.ThrowsAny<Exception>(() => Decompressor.InflatePrefix(deflated, 100));
    }

    [Fact]
    public void TheDecodedLengthLimitIs256MiB() {
        Assert.Equal(256 * 1024 * 1024, Decompressor.MAX_DECODED_LENGTH);
    }

    [Fact]
    public void InflateRejectsDataThatDecodesToMoreThanTheLimit() {
        byte[] deflated = Compressor.Deflate(new byte[1000]);
        Assert.Equal(1000, Decompressor.Inflate(deflated, 1000).Length);
        Exception e = Assert.ThrowsAny<Exception>(() => Decompressor.Inflate(deflated, 999));
        Assert.Equal("Flate data decodes to more than 999 bytes", e.Message);
    }

    [Fact]
    public void InflatePrefixReturnsTheFirstBytesAndIgnoresTheRest() {
        byte[] data = Ascii("hello hello hello hello");
        byte[] deflated = Compressor.Deflate(data);
        Assert.Equal(data[..5], Decompressor.InflatePrefix(deflated, 5));
        Assert.Equal(data, Decompressor.InflatePrefix(deflated, 100));
        byte[] truncated = deflated[..6];
        Assert.ThrowsAny<Exception>(() => Decompressor.InflatePrefix(truncated, 20));
    }

    [Fact]
    public void LzwDecodeRejectsDataThatDecodesToMoreThanTheLimit() {
        byte[] encoded = Bytes(0x80, 0x0B, 0x60, 0x50, 0x22, 0x0C, 0x0C, 0x85, 0x01);
        Assert.Equal(10, Decompressor.LZWDecode(encoded, 10).Length);
        Exception e = Assert.ThrowsAny<Exception>(() => Decompressor.LZWDecode(encoded, 9));
        Assert.Equal("LZW data decodes to more than 9 bytes", e.Message);
    }

    [Fact]
    public void RunLengthDecodeRejectsDataThatDecodesToMoreThanTheLimit() {
        byte[] encoded = Bytes(2, 'a', 'b', 'c', 254, 'x', 128);
        Assert.Equal(6, Decompressor.RunLengthDecode(encoded, 6).Length);
        Exception e = Assert.ThrowsAny<Exception>(() => Decompressor.RunLengthDecode(encoded, 5));
        Assert.Equal("RunLength data decodes to more than 5 bytes", e.Message);
    }

    [Fact]
    public void DeflateOfNoBytesIsAnEmptyZlibStream() {
        byte[] deflated = Compressor.Deflate(new byte[0]);
        Assert.Equal(8, deflated.Length);
        Assert.Equal(0x78, deflated[0]);
    }

    [Fact]
    public void InflateAcceptsAnEmptyStream() {
        Assert.Empty(Decompressor.Inflate(Compressor.Deflate(new byte[0])));
    }
}
}

File: tests/java/com/pdfjet/DecompressorTest.java
/*
 * DecompressorTest.java
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
package com.pdfjet;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Random;
import org.junit.jupiter.api.Test;

/** The stream filters and predictors that reading a PDF needs, and the compressor. */
class DecompressorTest {
    private static byte[] bytes(int... values) {
        byte[] result = new byte[values.length];
        for (int i = 0; i < values.length; i++) {
            result[i] = (byte) values[i];
        }
        return result;
    }

    private static byte[] ascii(String text) {
        return text.getBytes(StandardCharsets.US_ASCII);
    }

    @Test
    void lzwDecodesTheExampleOfTheStandard() throws Exception {
        // ISO 32000-1, 7.4.4.2: the encoding of "-----A---B".
        byte[] encoded = bytes(0x80, 0x0B, 0x60, 0x50, 0x22, 0x0C, 0x0C, 0x85, 0x01);
        assertEquals("-----A---B", TestSupport.latin1(Decompressor.lzwDecode(encoded)));
    }

    @Test
    void asciiHexDecodeSkipsWhitespaceAndStopsAtTheEndMarker() {
        assertEquals("Hello", TestSupport.latin1(Decompressor.asciiHexDecode(ascii("48 65 6C6C6F>"))));
    }

    @Test
    void asciiHexDecodeReadsAMissingLastDigitAsZero() {
        assertArrayEquals(bytes(0x70), Decompressor.asciiHexDecode(ascii("7>")));
    }

    @Test
    void ascii85DecodesWithAndWithoutThePrefix() {
        // base64.a85encode(b"Hello world") in Python.
        assertEquals("Hello world", TestSupport.latin1(Decompressor.ascii85Decode(ascii("87cURD]j7BEbo7~>"))));
        assertEquals("Hello world", TestSupport.latin1(Decompressor.ascii85Decode(ascii("<~87cURD]j7BEbo7~>"))));
    }

    @Test
    void ascii85DecodesZAsFourZeroBytesAndAPartialLastGroup() {
        assertArrayEquals(bytes(0, 0, 0, 0), Decompressor.ascii85Decode(ascii("z~>")));
        assertArrayEquals(bytes(0, 0, 0, 0, 'a', 'b'), Decompressor.ascii85Decode(ascii("z@:B~>")));
    }

    @Test
    void runLengthDecodeCopiesLiteralsAndRepeatsRuns() throws Exception {
        byte[] encoded = bytes(2, 'a', 'b', 'c', 254, 'x', 128);
        assertEquals("abcxxx", TestSupport.latin1(Decompressor.runLengthDecode(encoded)));
    }

    @Test
    void pngPredictorsUndoEachRowFilter() {
        // Each row is the filter type and three one byte samples.
        assertArrayEquals(bytes(1, 2, 3), Decompressor.applyPredictor(bytes(1, 1, 1, 1), 11, 1, 8, 3));
        assertArrayEquals(bytes(1, 2, 3, 2, 3, 4), Decompressor.applyPredictor(bytes(0, 1, 2, 3, 2, 1, 1, 1), 12, 1, 8, 3));
        assertArrayEquals(bytes(2, 5, 8), Decompressor.applyPredictor(bytes(3, 2, 4, 6), 13, 1, 8, 3));
        assertArrayEquals(bytes(1, 2, 3), Decompressor.applyPredictor(bytes(4, 1, 1, 1), 14, 1, 8, 3));
    }

    @Test
    void tiffPredictorAddsTheSampleToTheLeft() {
        assertArrayEquals(bytes(1, 2, 3, 5, 5, 5), Decompressor.applyPredictor(bytes(1, 1, 1, 5, 0, 0), 2, 1, 8, 3));
    }

    @Test
    void predictorOneAndInvalidParametersLeaveTheDataAsItIs() {
        byte[] data = bytes(9, 8, 7);
        assertArrayEquals(data, Decompressor.applyPredictor(data, 1, 1, 8, 3));
        assertArrayEquals(data, Decompressor.applyPredictor(data, 12, 0, 8, 3));
    }

    @Test
    void inflateUndoesDeflate() throws Exception {
        byte[] data = new byte[100000];
        new Random(42).nextBytes(data);
        Arrays.fill(data, 50000, 90000, (byte) 'x');
        assertArrayEquals(data, Decompressor.inflate(Compressor.deflate(data)));
    }

    @Test
    void inflateRejectsATruncatedStream() {
        byte[] deflated = Compressor.deflate(ascii("hello hello hello hello"));
        assertThrows(Exception.class, () -> Decompressor.inflate(Arrays.copyOf(deflated, 6)));
    }

    @Test
    void inflateIgnoresBytesAfterTheEndOfAStream() throws Exception {
        byte[] data = ascii("hello hello hello hello");
        byte[] deflated = Compressor.deflate(data);
        byte[] padded = Arrays.copyOf(deflated, deflated.length + 3);
        padded[deflated.length] = '\r';
        padded[deflated.length + 1] = '\n';
        padded[deflated.length + 2] = 0x42;
        assertArrayEquals(data, Decompressor.inflate(padded));
    }

    @Test
    void inflateRejectsEveryTruncationOfAStream() throws Exception {
        byte[] data = new byte[2000];
        new Random(3).nextBytes(data);
        Arrays.fill(data, 1000, 2000, (byte) 'a');
        byte[] deflated = Compressor.deflate(data);
        assertArrayEquals(data, Decompressor.inflate(deflated));
        // The last 4 bytes are the Adler-32 checksum, which is checked too.
        for (int length = 0; length < deflated.length; length++) {
            final byte[] truncated = Arrays.copyOf(deflated, length);
            assertThrows(Exception.class, () -> Decompressor.inflate(truncated), "length " + length);
        }
    }

    @Test
    void inflateRejectsAWrongChecksumOrHeader() throws Exception {
        byte[] deflated = Compressor.deflate("PDFjet PDFjet PDFjet".getBytes(StandardCharsets.US_ASCII));
        final byte[] wrongChecksum = deflated.clone();
        wrongChecksum[wrongChecksum.length - 1] ^= 1;
        assertThrows(Exception.class, () -> Decompressor.inflate(wrongChecksum));
        final byte[] wrongMethod = deflated.clone();
        wrongMethod[0] = 0x77;     // Method 7, and the check no longer fits
        assertThrows(Exception.class, () -> Decompressor.inflate(wrongMethod));
        final byte[] wrongCheck = deflated.clone();
        wrongCheck[1] ^= 1;
        assertThrows(Exception.class, () -> Decompressor.inflate(wrongCheck));
    }

    @Test
    void inflatePrefixWithItsBytesNeedsNoChecksum() throws Exception {
        byte[] data = "PDFjet PDFjet PDFjet".getBytes(StandardCharsets.US_ASCII);
        final byte[] deflated = Compressor.deflate(data);
        deflated[deflated.length - 1] ^= 1;
        assertArrayEquals(Arrays.copyOf(data, 5), Decompressor.inflatePrefix(deflated, 5));
        assertThrows(Exception.class, () -> Decompressor.inflatePrefix(deflated, 100));
    }

    @Test
    void theDecodedLengthLimitIs256MiB() {
        assertEquals(256 * 1024 * 1024, Decompressor.MAX_DECODED_LENGTH);
    }

    @Test
    void inflateRejectsDataThatDecodesToMoreThanTheLimit() throws Exception {
        final byte[] deflated = Compressor.deflate(new byte[1000]);
        assertEquals(1000, Decompressor.inflate(deflated, 1000).length);
        Exception e = assertThrows(Exception.class, () -> Decompressor.inflate(deflated, 999));
        assertEquals("Flate data decodes to more than 999 bytes", e.getMessage());
    }

    @Test
    void inflatePrefixReturnsTheFirstBytesAndIgnoresTheRest() throws Exception {
        byte[] data = ascii("hello hello hello hello");
        final byte[] deflated = Compressor.deflate(data);
        assertArrayEquals(Arrays.copyOf(data, 5), Decompressor.inflatePrefix(deflated, 5));
        assertArrayEquals(data, Decompressor.inflatePrefix(deflated, 100));
        assertThrows(Exception.class, () -> Decompressor.inflatePrefix(Arrays.copyOf(deflated, 6), 20));
    }

    @Test
    void lzwDecodeRejectsDataThatDecodesToMoreThanTheLimit() throws Exception {
        final byte[] encoded = bytes(0x80, 0x0B, 0x60, 0x50, 0x22, 0x0C, 0x0C, 0x85, 0x01);
        assertEquals(10, Decompressor.lzwDecode(encoded, 10).length);
        Exception e = assertThrows(Exception.class, () -> Decompressor.lzwDecode(encoded, 9));
        assertEquals("LZW data decodes to more than 9 bytes", e.getMessage());
    }

    @Test
    void runLengthDecodeRejectsDataThatDecodesToMoreThanTheLimit() throws Exception {
        final byte[] encoded = bytes(2, 'a', 'b', 'c', 254, 'x', 128);
        assertEquals(6, Decompressor.runLengthDecode(encoded, 6).length);
        Exception e = assertThrows(Exception.class, () -> Decompressor.runLengthDecode(encoded, 5));
        assertEquals("RunLength data decodes to more than 5 bytes", e.getMessage());
    }

    @Test
    void deflateOfNoBytesIsAnEmptyZlibStream() {
        byte[] deflated = Compressor.deflate(new byte[0]);
        assertEquals(8, deflated.length);
        assertEquals(0x78, deflated[0] & 0xFF);
    }

    @Test
    void inflateAcceptsAnEmptyStream() throws Exception {
        assertEquals(0, Decompressor.inflate(Compressor.deflate(new byte[0])).length);
    }
}
